lib/craft/machine.ex

```diff
@@ -344,7 +379,30 @@ defmodule Craft.Machine do
   def handle_call({:query, :eventual, query}, from, state) do
     Logger.debug("executing query", logger_metadata(trace: {:query, :eventual, :quorum_read, from, query}))
 
-    {:reply, state.module.handle_query(query, state.private), state}
+    case state.module.handle_query(query, {:direct, from}, state.private) do
+      {:reply, reply} -> {:reply, reply, state}
+      :noreply -> {:noreply, state}
+    end
+  end
+
+  def handle_call({{:query_reply, reply}, query_from}, _from, state) do
+    is_pending? = MapSet.member?(state.pending_parallel_queries, query_from)
+    quorum_happend? = !is_nil(state.last_quorum_at) and :erlang.monotonic_time(:millisecond) > state.last_quorum_at
```

Comment:
unfortunately, this won't work, the time at which the query respond will always be greater than the time of the last quorum. 

the query time (i.e. when the snapshot for the query was taken) should happen before query execution and be handed opaquely to the application machine's callback, and then handed back to this machine in the `:query_reply` tuple.